cmake/nuttx/px4_impl_nuttx.cmake

```diff
@@ -248,43 +248,48 @@ function(px4_nuttx_add_export)
 		REQUIRED OUT CONFIG THREADS
 		ARGN ${ARGN})
 
-	set(nuttx_src ${PX4_BINARY_DIR}/${CONFIG}/NuttX)
+	set(nuttx_build_src ${PX4_BINARY_DIR}/${CONFIG}/NuttX)
+	set(nuttx_export_dir ${nuttx_build_src}/nuttx/nuttx-export)
 
-	# all patches
+	# nuttx cmake dependency files
+	set(nuttx_copy ${PX4_BINARY_DIR}/${CONFIG}_nuttx_copy.stamp)
+	set(nuttx_configure ${PX4_BINARY_DIR}/${CONFIG}_nuttx_configure.stamp)
+	set(nuttx_export ${PX4_BINARY_DIR}/${CONFIG}_nuttx_export.stamp)
+
+	# all nuttx patches
 	file(GLOB nuttx_patches ${PX4_SOURCE_DIR}/nuttx-patches/*.patch)
 	list(SORT nuttx_patches)
 
 	# copy
-	file(GLOB_RECURSE nuttx_all_files ${PX4_SOURCE_DIR}/NuttX/*)
+	file(GLOB_RECURSE nuttx_all_files ${PX4_SOURCE_DIR}/NuttX/nuttx/Makefile)
 	file(RELATIVE_PATH nuttx_cp_src ${PX4_BINARY_DIR} ${PX4_SOURCE_DIR}/NuttX)
-	add_custom_command(OUTPUT ${PX4_BINARY_DIR}/nuttx_copy_${CONFIG}.stamp
-		COMMAND ${MKDIR} -p ${nuttx_src}
-		COMMAND rsync -a --delete --exclude=.git ${nuttx_cp_src}/ ${CONFIG}/NuttX/
-		COMMAND ${TOUCH} ${PX4_BINARY_DIR}/nuttx_copy_${CONFIG}.stamp
-		DEPENDS ${DEPENDS} ${nuttx_patches} ${nuttx_all_files}
+	add_custom_command(OUTPUT ${nuttx_copy}
+		COMMAND ${MKDIR} -p ${nuttx_build_src}
+		COMMAND rsync -rp --delete --exclude=.git --exclude=nuttx-export ${nuttx_cp_src}/ ${CONFIG}/NuttX/
+		COMMAND cmake -E touch ${nuttx_copy}
```

Comment:
Are you talking about top level NuttX, or the copied and patched version? If you touch anything top level it rsyncs everything and then it needs to apply all patches, full and rebuild everything. We would be able to do it per file, but we don't know what the patches actually modify. We could try and get fancy and test apply, but that seems like a convoluted approach.

If we want to do any better I'd look for a way to continue keeping the patches in tree, but apply them as commits to the NuttX submodule as the first step of the build. Then the Nuttx copy per PX4 build could actually be hard links back to the source and you could debug, edit, rebuild in place.

There are git tools for moving between patches <-> git commits, but I haven't really used them.